### DISCUSSION TITLE
interface Options are not recognised in the code after migration from v6 to v7

### DISCUSSION THREAD

**Discussion by denis-a1qa** [webdriverio/codemod#13] [Migration Issue]
Hello!

In v6 we were using following interfaces to expand webdriverio (it might be incorrect but it was working fine across project):
```/**
 * Extensible interface for adding custom commands or properties, for global visibility.
 * Added commands must be implemented in all webdriver versions.
 */
// tslint:disable-next-line: no-namespace
declare namespace WebdriverIO {
    /**
     * Interface for adding custom commands
     */
    interface Browser {
        // Custom commands for `browser`
        setLocationService(condition: boolean): void;
        setWiFi(condition: boolean): void;
    }

    /**
     * Interface for adding custom options for BrowserObject.config (browser.config.customProrerty)
     */
    interface Options {
        testType?: string;
        logElementsActions?: boolean;
    }
}

// tslint:disable-next-line: no-namespace
declare namespace WebDriver {
    /**
     * Interface for adding custom porperties for BrowserObject (browser.customProperty)
     */
    interface ClientOptions {
        isNativeTest?: boolean;
        logElementsActions?: boolean;
    }
}
```

we were able to access config properties anywhere without any typescript errors (autocomplete was also working fine)

```
private addCustomProperties() {
        driver.isNativeTest = driver.config.testType === 'native';
        driver.logElementsActions = driver.config.logElementsActions;
    }
```

But after upgrading to v7 and chnaing code to the following one:

```
export { };
/**
 * Extensible interface for adding custom commands or properties, for global visibility.
 * Added commands must be implemented in all webdriver versions.
 */
declare global {
    namespace WebdriverIO {
        /**
         * Interface for adding custom commands
         */
        interface Browser {
            // Custom commands for `browser`
            setLocationService(condition: boolean): void;
            setWiFi(condition: boolean): void;
            isNativeTest?: boolean;
            logElementsActions?: boolean;
        }

        interface Options {
            testType?: string;
            logElementsActions?: boolean;
        }
    }

}
```
We no longer see custom config properties and need to use "hacks" with any:

```
/**
 * Base class for implementing the addition of custom commands and properties that are missing.
 * This is an abstraction for Mobile and Desktop webdrivers.
    }

    private addCustomProperties() {
        driver.isNativeTest = (driver.config as any).testType === 'native';
        driver.logElementsActions = (driver.config as any).logElementsActions;
    }

    /**
```

Any ideas how to fix that? It doesn't affect test executions but I really hate using `as any`

**Comment by denis-a1qa** [comment]
I must say that i didn't use codemod and was doing everything manually (there is not that many files that need to be changed).

**Comment by osmolyar** [comment]
@denis-a1qa , this is because using "declare global" in the type definition file causes it to be recognized as a "module" rather than an ambient type definition file by typescript.  If you are using ambient type definition files, e.g., *.d.ts files anywhere in the project that get picked up as type definition files without specifying them in includes/files, then you can't have any export, import or declare global statements in them.

The solution is simply to change the declaration to 

`declare namespace WebdriverIO`

(which it sounds like you had before; some people were using `declare module WebdriverIO` with v6 or earlier.)  Anyway `declare namespace WebdriverIO` works fine with v7.

I believe the documentation was updated to make the distinction between module and ambient type def files as part of https://github.com/webdriverio/webdriverio/pull/6719.

**Reply by denis-a1qa** [reply]
@osmolyar Thank you for reponse. My issue is not with custom methods, but rather with custom options for config that can be accessed like this:

`driver.config.<option>`

in v6 i could add such options using:
```
declare namespace WebdriverIO {
    /**
     * Interface for adding custom options for BrowserObject.config (browser.config.customProrerty)
     */
    interface Options {
        testType?: string;
        logElementsActions?: boolean;
    }
}
```

But in v7 I can't add such options and can only hide TS errors with any
`(driver.config as any).<option>`

**Reply by christian-bromann** [reply]
@denis-a1qa can you try:
```ts
declare global {
    namespace WebdriverIO {
        interface Config {
            testType?: string;
            logElementsActions?: boolean;
        }
    }
}
```

Generally I believe we like to move away assigning custom options to the wdio config. I don't think they belong in there. You can store them in a custom page object or even globally.
